### PR TITLE
feat: Migrate /ksql and /ksql/terminate endpoints

### PR DESCRIPTION
**Pull request by purplefox**
### Description 

This PR migrates the /ksql and /ksql/terminate endpoint to the new Vert.x implementation.

It also includes some changes around security to make sure things work properly with the custom security plugins in CP and Cloud.

It also some refactoring and cleanup in the verticle handlers.

Once this PR is merged it means the /ksql and /ksql/terminate endpoints will no longer be served by Jetty. This means we've had to also introduce new custom authentication plugins for CP and cloud, and those must be merged before we merge this.

### Testing done 

Non functional change.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: hs_err_pid47193.log [hunk 1]
```diff
@@ -0,0 +1,612 @@
+#
```
Comment (big-andy-coates) [inline hs_err_pid47193.log]: I'm assuming committing this was unintentional ;)

Probably worth updating git ignore...

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ */
+public class AuthenticationPluginHandler implements Handler<RoutingContext> {
+
+  private final Server server;
+  private final AuthenticationPlugin securityHandlerPlugin;
+
+  public AuthenticationPluginHandler(final Server server,
+      final AuthenticationPlugin securityHandlerPlugin) {
+    this.server = server;
+    this.securityHandlerPlugin = securityHandlerPlugin;
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java]: nit: validate params that will be stored in object state; ensuring object does not get into an invalid state.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      return null;
+    });
+  }
+
+  private static class AuthPluginUser implements ApiUser {
+
+    private final Principal principal;
+
+    AuthPluginUser(final Principal principal) {
+      this.principal = principal;
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java]: nit: validate params that will be stored in object state; ensuring object does not get into an invalid state.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/JaasPrincipal.java [hunk 4]
```diff
@@ -17,11 +17,14 @@
...
-class ApiPrincipal implements Principal {
+/**
+ * Principal implementation created when authenticating with the JaasAuthProvider
+ */
+class JaasPrincipal implements Principal {
 
   private final String name;
 
-  ApiPrincipal(final String name) {
+  JaasPrincipal(final String name) {
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/JaasPrincipal.java]: nit: validate params that will be stored in object state; ensuring object does not get into an invalid state.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/KsqlAuthorizationProviderHandler.java [hunk 5]
```diff
@@ -23,9 +23,12 @@
 import io.vertx.core.WorkerExecutor;
 import io.vertx.ext.auth.User;
 import io.vertx.ext.web.RoutingContext;
-import java.security.Principal;
 import java.util.Set;
 
+/**
+ * Handler that calls a KsqlAuthorizationProvider plugin that can be used for custom authorization
+ * (e.g. RBAC in Confluent platform)
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/KsqlAuthorizationProviderHandler.java]: We generally avoid reference to RBAC / CP in the OSS ksqlDB code base.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/endpoints/KsqlServerEndpoints.java [hunk 6]
```diff
@@ -42,17 +46,22 @@
...
       final PullQueryExecutor pullQueryExecutor,
-      final KsqlSecurityContextProvider ksqlSecurityContextProvider) {
+      final KsqlSecurityContextProvider ksqlSecurityContextProvider,
+      final KsqlResource ksqlResource) {
     this.ksqlEngine = Objects.requireNonNull(ksqlEngine);
     this.ksqlConfig = Objects.requireNonNull(ksqlConfig);
     this.pullQueryExecutor = Objects.requireNonNull(pullQueryExecutor);
     this.reservedInternalTopics = new ReservedInternalTopics(ksqlConfig);
     this.ksqlSecurityContextProvider = Objects.requireNonNull(ksqlSecurityContextProvider);
+    this.ksqlStatementsEndpoint = new KsqlStatementsEndpoint(Objects.requireNonNull(ksqlResource));
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/endpoints/KsqlServerEndpoints.java]: nit: no need to null check `ksqlResource` here - that should be the responsibility of the `KsqlStatementsEndpoint` constructor...

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/endpoints/KsqlStatementsEndpoint.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import io.confluent.ksql.security.KsqlSecurityContext;
+import io.confluent.ksql.util.VertxUtils;
+import javax.ws.rs.core.Response;
+
+public class KsqlStatementsEndpoint {
+
+  private final KsqlResource ksqlResource;
+
+  public KsqlStatementsEndpoint(final KsqlResource ksqlResource) {
+    this.ksqlResource = ksqlResource;
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/endpoints/KsqlStatementsEndpoint.java]: nit: validate params that will be stored in object state; ensuring object does not get into an invalid state.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/endpoints/TerminateEndpoint.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import io.confluent.ksql.security.KsqlSecurityContext;
+import io.confluent.ksql.util.VertxUtils;
+import javax.ws.rs.core.Response;
+
+public class TerminateEndpoint {
+
+  private final KsqlResource ksqlResource;
+
+  public TerminateEndpoint(final KsqlResource ksqlResource) {
+    this.ksqlResource = ksqlResource;
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/endpoints/TerminateEndpoint.java]: nit: validate params that will be stored in object state; ensuring object does not get into an invalid state.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/InsertsStreamHandler.java [hunk 9]
```diff
@@ -157,9 +157,10 @@ private Void handleInsertSubscriberException(final Throwable t,
       }
       log.error("Failed to execute inserts", toLog);
       // We don't expose internal error message via public API
-      ServerUtils.handleError(routingContext.response(), 500, ErrorCodes.ERROR_CODE_INTERNAL_ERROR,
-          "The server encountered an internal error when processing inserts."
-              + " Please consult the server logs for more information.");
+      routingContext.fail(500,
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/InsertsStreamHandler.java]: Can we use one of the enums for more descriptive status codes please, e.g. `HttpStatusCode.SERVER_ERROR` rather than `500`.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PortedEndpoints.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+public class PortedEndpoints {
+
+  private static final Set<String> PORTED_ENDPOINTS = ImmutableSet.of("/ksql");
+
+  private final Endpoints endpoints;
+  private final Server server;
+
+  public PortedEndpoints(final Endpoints endpoints, final Server server) {
+    this.endpoints = endpoints;
+    this.server = server;
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PortedEndpoints.java]: nit: validate params that will be stored in object state; ensuring object does not get into an invalid state.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/QueryStreamHandler.java [hunk 11]
```diff
@@ -114,19 +113,19 @@ private Void handleQueryPublisherException(final Throwable t,
     if (t instanceof CompletionException) {
       final Throwable actual = t.getCause();
       if (actual instanceof KsqlStatementException) {
-        ServerUtils.handleError(routingContext.response(), 400, ErrorCodes.ERROR_CODE_INVALID_QUERY,
-            actual.getMessage());
+        routingContext.fail(400,
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/QueryStreamHandler.java]: ditto on the enum vs `400`.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/QueryStreamHandler.java [hunk 12]
```diff
@@ -114,19 +113,19 @@ private Void handleQueryPublisherException(final Throwable t,
...
-        ServerUtils
-            .handleError(routingContext.response(), 400, ((KsqlApiException) actual).getErrorCode(),
-                actual.getMessage());
+        routingContext.fail(400, actual);
+        return null;
       }
     }
     // We don't expose internal error message via public API
-    ServerUtils.handleError(routingContext.response(), 500, ErrorCodes.ERROR_CODE_INTERNAL_ERROR,
+    routingContext.fail(500, new KsqlApiException(
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/QueryStreamHandler.java]: ditto on the enum vs `500`

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java [hunk 13]
```diff
@@ -34,47 +35,36 @@
...
 
   private static class HttpResponseErrorHandler implements PojoDeserializerErrorHandler {
 
-    private final HttpServerResponse response;
+    private final RoutingContext routingContext;
 
-    HttpResponseErrorHandler(final HttpServerResponse response) {
-      this.response = response;
+    HttpResponseErrorHandler(final RoutingContext routingContext) {
+      this.routingContext = routingContext;
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java]: nit: validate params that will be stored in object state; ensuring object does not get into an invalid state.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/spi/EndpointResponse.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+public interface EndpointResponse {
+
+  int getStatusCode();
+
+  String getStatusMessage();
+
+  Object getResponseBody();
+
+  static EndpointResponse create(final int errorCode, final String statusMessage,
+      final Object responseBody) {
```
Comment (big-andy-coates) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/spi/EndpointResponse.java]: nit: validate params that will be stored in object state; ensuring object does not get into an invalid state.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/InsertsStreamHandler.java (hunk 9 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/InsertsStreamHandler.java]: Imho, I prefer explicit error codes rather than indirecting through _yet another 3rd party dependency_. Constants files are great for constants that might change in the future, but HTTP status codes are never going to change.
But.. in the interests of an easy life, I don't mind changing it.
We can't use the Jetty error code constants as Jetty is going away.
I guess we could org.apache.HttpStatus which is currently being dragged in anyway.

File: hs_err_pid47193.log (hunk 1 above)
Comment (purplefox) [reply hs_err_pid47193.log]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java (hunk 2 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java (hunk 3 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/JaasPrincipal.java (hunk 4 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/JaasPrincipal.java]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/KsqlAuthorizationProviderHandler.java (hunk 5 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/KsqlAuthorizationProviderHandler.java]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/endpoints/KsqlServerEndpoints.java (hunk 6 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/endpoints/KsqlServerEndpoints.java]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/endpoints/KsqlStatementsEndpoint.java (hunk 7 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/endpoints/KsqlStatementsEndpoint.java]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/endpoints/TerminateEndpoint.java (hunk 8 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/endpoints/TerminateEndpoint.java]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PortedEndpoints.java (hunk 10 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PortedEndpoints.java]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/QueryStreamHandler.java (hunk 11 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/QueryStreamHandler.java]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/QueryStreamHandler.java (hunk 12 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/QueryStreamHandler.java]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java (hunk 13 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/spi/EndpointResponse.java (hunk 14 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/spi/EndpointResponse.java]: ack

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+      routingContext.next();
+      return;
+    }
+    final CompletableFuture<Principal> cf = securityHandlerPlugin
+        .handleAuth(routingContext, server.getWorkerExecutor());
+    cf.thenAccept(principal -> {
+      if (principal == null) {
+        // Not authenticated
+        routingContext
+            .fail(HttpStatus.SC_UNAUTHORIZED, new KsqlApiException("Failed authentication",
```
Comment (vcrfxia) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java]: Why is the plugin handler failing the context? This seems to contradict the expectation that implementations of `AuthenticationPlugin` should end the response if the user fails to authenticate.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java [hunk 16]
```diff
@@ -34,47 +37,39 @@
   private ServerUtils() {
   }
 
-  public static void handleError(final HttpServerResponse response, final int statusCode,
-      final int errorCode, final String errMsg) {
-    final ErrorResponse errorResponse = new ErrorResponse(errorCode, errMsg);
-    final Buffer buffer = errorResponse.toBuffer();
-    response.setStatusCode(statusCode).end(buffer);
```
Comment (vcrfxia) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java]: Before this refactor, `handleError()` was populating the status code and providing a message when ending the response, but the new code that previously called this method does not. Was this a conscious decision? I think it's nice to populate the response because otherwise failures appear silent from curl (and probably other tools as well?).

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PortedEndpoints.java [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.handler.BodyHandler;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiFunction;
+import javax.ws.rs.core.MediaType;
+import org.apache.http.HttpStatus;
+
+public class PortedEndpoints {
```
Comment (vcrfxia) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PortedEndpoints.java]: Why is this public?

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PortedEndpoints.java [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+          .setStatusMessage(endpointResponse.getStatusMessage())
+          .end(responseBody);
+
+    }).exceptionally(t -> {
+      routingContext.fail(HttpStatus.SC_INTERNAL_SERVER_ERROR, t);
+      return null;
+    });
+  }
+
+  public static void oldApiFailureHandler(final RoutingContext routingContext) {
```
Comment (vcrfxia) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PortedEndpoints.java]: Why is this public?

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java [hunk 19]
```diff
@@ -34,47 +37,39 @@
...
+              ERROR_CODE_MISSING_PARAM));
     }
 
     @Override
     public void onExtraParam(final String paramName) {
-      handleError(response, 400, ErrorCodes.ERROR_CODE_UNKNOWN_PARAM,
-          "Unknown arg " + paramName);
+      routingContext
+          .fail(HttpStatus.SC_BAD_REQUEST, new KsqlApiException("Unknown arg " + paramName,
+              ErrorCodes.ERROR_CODE_UNKNOWN_PARAM));
```
Comment (vcrfxia) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java]: nit: inconsistent static imports. `ERROR_CODE_MISSING_PARAM` above is a static import but this error code (and the one below) aren't.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerVerticle.java [hunk 20]
```diff
@@ -52,8 +54,11 @@
   // CHECKSTYLE_RULES.ON: ClassDataAbstractionCoupling
   private static final Logger log = LoggerFactory.getLogger(ServerVerticle.class);
 
-  private static final Set<String> NEW_API_ENDPOINTS = ImmutableSet
-      .of("/query-stream", "/inserts-stream", "/close-query");
+  private static final Set<String> NON_PROXIED_ENDPOINTS = ImmutableSet
+      .of("/query-stream", "/inserts-stream", "/close-query", "/ksql");
+
+  // Quick switch so we can easily revert to not serving ported endpoints directly
+  private static final boolean SERVE_PORTED_ENDPOINTS = true;
```
Comment (vcrfxia) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerVerticle.java]: Thanks for including this. Hopefully we won't need it but it's reassuring to know we can easily undo this change if something comes up.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java (hunk 15 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java]: Previously the plugin was ending the response, which seemed wrong to me (as per our previous discussion) so I changed the contract, so now it's Vert.x that does this.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java (hunk 16 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java]: Not sure I follow.. could you elaborate a bit?

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PortedEndpoints.java (hunk 17 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PortedEndpoints.java]: No particular reason, I'm usually make classes public by default unless there's a very good reason not to.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PortedEndpoints.java (hunk 18 above)
Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PortedEndpoints.java]: As above

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java [hunk 21]
```diff
@@ -234,6 +236,10 @@ public static KsqlRestConfig convertToApiServerConfig(final KsqlRestConfig confi
           origs.put(ApiServerConfig.TLS_CLIENT_AUTH_REQUIRED, "required");
         }
       }
+      final List<String> unauthedPaths = config.getList(RestConfig.AUTHENTICATION_SKIP_PATHS);
+      if (unauthedPaths != null) {
+        origs.put(ApiServerConfig.AUTHENTICATION_SKIP_PATHS_CONFIG, unauthedPaths);
+      }
```
Comment (spena) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java]: Seems this is set only if `keyStoreLocation` is not null/empty. Why is that?

Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java]: Doh!

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java (hunk 15 above)
Comment (vcrfxia) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/auth/AuthenticationPluginHandler.java]: Can we update the Java docs in `AuthenticationPlugin` and the test implementation (in this repo) to reflect that? The fact that the two were incongruous threw me off.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java (hunk 16 above)
Comment (vcrfxia) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java]: The old code (the line I've commented on here) did:
```
response.setStatusCode(statusCode).end(buffer);
```
whereas the new code does:
```
routingContext.fail(statusCode, exception);
```
The latter doesn't populate the response status code and message (buffer), right? As a result, failed curl requests come up blank. I remember being really confused by this when doing manual testing.

Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java]: Could you point me to the routingContext.fail() line you are referring to? There are probably a few places in the code that call that :)

RoutingContext fail will cause the failureHandler to be called.

Comment (vcrfxia) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java]: Ah I see, and both the regular failure handler (in `ServerVerticle`) and the old API failure handler (in `PortedEndpoints`) set the status code and message, so we should be good. Thanks for clarifying!

Comment (purplefox) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerUtils.java]: Right, the change here basically centralises all the failure handling in one place so we don't end responses with failure from multiple places any more, which makes it a bit simpler to manage.

**Comment by purplefox** [comment]
>> IIUC, once this PR is merged the old configs become no-ops, which seems dangerous as users might not realize

In this PR it's only the /ksql/* endpoints that are being migrated, the other ones remain hosted by an internal netty. This means that both the old Jetty plugin classes AND the new Vert.x ones need to be specified if someone configures ksql for RBAC (or cloud).

This will be the case until all the endpoints are migrated at which point only the new plugin will be needed.

I have a note to update the docs about configuring RBAC to include the new plugin.

**Comment by purplefox** [comment]
>> We should manually verify that the migrated security plugins work with this change. I realize this suite of changes has already been significantly delayed as a result of trying to do this but I think it's important to do our due diligence with security changes like these. I think it's fine to say you don't want to be the person who performs the manual verification but we should make sure it gets done.

Agreed. I already tested this with RBAC, it's just cloud that needs doing. Any volunteers? :)

**Comment by vcrfxia** [comment]
> In this PR it's only the /ksql/* endpoints that are being migrated, the other ones remain hosted by an internal Jetty. This means that both the old Jetty plugin classes AND the new Vert.x ones need to be specified if someone configures ksql for RBAC (or cloud).
> 
> This will be the case until all the endpoints are migrated at which point only the new plugin will be needed.

Makes sense. What do you think about throwing an exception if the Jetty configs are supplied but not the new vert.x ones, during the transition period? I'm worried about leaving master in a partial state now that we cut releases directly from master.

**Comment by purplefox** [comment]
> Makes sense. What do you think about throwing an exception if the Jetty configs are supplied but not the new vert.x ones, during the transition period?

Ack. I've added a check.
